### PR TITLE
addition of fixed and dynamic time series by double clicking on map

### PR DESCRIPTION
**Pull request by wzhengui**
Hi @pmav99 and @brey : 

I finished adding the time series. The function seems working now for both fixed stations and dynamic location (mouse point). Only minor issue is that you need to refresh the figures after removing stations. 

Could you help double check? If you think OK, please merge it to the master.  

Thanks,
Zhengui

### CODE REVIEW

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py]: if we pass `fid` (i.e. an instance of `ThalassaUI`) as an argument to `get_timeseries()` we practically couple the User Interface with the API. I think that we should avoid that. 

Instead, we should try to design the API functions in such a way that they can be used on their own. The goal is to be able to use them without even knowing that `ThalassaUI` even exists. 

I haven't run this, so it is probably not running as is, but the general idea would be to use something like this:

``` python
def get_timestamp(trimesh: gv.Trimesh, add_points: bool = True, remove_points: bool = True): 
    ...
```

**Review by pmav99 — CHANGES_REQUESTED** [review]
Thank you @wzhengui, overall it seems like a great step forward. Some things need to be taken care of before we can merge this, though:

## Y axis does not get updated

The main issue that I see is that In the timeseries plots the scale of the Y axis is not being updated which results in plots like the following:

image

If possible, we should update the Y axis each time a new point is added or the location of the mouse has changed. If that is not possible, then the plot should be created using the global Min/Max values 

## Plots over land

Hovering the mouse over land is still creating a timeseries plot. I understand that when we are close to the coast, it might be more difficult to infer whether we should create the plot or not,  but, ideally, hovering the mouse over e.g. Switzerland should not create a plot. 

## Legend label for current mouse point

If possible, I think it would be nice if the legend of the timeseries plot displayed a label for the current mouse location (i.e. the dotted line).

## Point removal

When the point removal action is active, double clicking removes the last added point regardless of the location of the mouse. This is not necessarily a deal breaker but it feels a bit counter-intuitive. Ideally we should remove points by clicking on them. If this is not feasible, then perhaps removing all points when the UI button is pressed could be considered. Nevertheless, I wouldn't block merging over this.

## Splitting up `get_timeseries()`

Currently we have a single function that extracts the timeseries, adds/removes points from the main plot and creates the timeseries plot. At the very least, the timeseries extraction should be a separate function.

## Code style

 Before getting this merged the coding style will need to be fixed. This is mostly my bad, since I didn't include instructions or setup linters with github actions. I will try to set them up later today. To help you get going, we will be following PEP8 (for the most part) and we will be using [black](https://github.com/psf/black) for formatting (with 109 characters as the line length).

1. Please, let's use just a single statement per line. E.g.
``` diff
-        dist=abs(sxi+1j*syi-x-1j*y);  nid=np.nonzero(dist==dist.min())[0][0]
-        mti=fdata['time'].data;       myi=fdata['elev'].data[:,nid].copy()
+        dist=abs(sxi+1j*syi-x-1j*y);
+        nid=np.nonzero(dist==dist.min())[0][0]
+        mti=fdata['time'].data;
+        myi=fdata['elev'].data[:,nid].copy()
```
2. No single-line ifs
3. Let's use descriptive variable names. Unless it is completely clear what they mean, names like `fid`, `hp`, `hs` etc should be avoided

**Comment by wzhengui** [comment]
Hi @pmav99 : I see some of the changes can be easily made, while some may take some more time in the testing. I will work on them when I find time. I think the most important thing is that the overall function is there, and these minor issues can be tackled in the iteration of code development.

**Comment by wzhengui** [comment]
Hi Panos:

Do you know how to set the extent of gv.WMTS (e.g.
xlim=(0,30),ylim=(0,30))? I searched for a while, but didn't make it work
in Thalassa.

Thanks,
Zhengui

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:11]: Please, one import per line: https://stackoverflow.com/questions/24271961/is-there-anything-bad-about-having-multiple-imports-on-one-line/25027300

File: thalassa/ui.py
Comment (pmav99) [inline thalassa/ui.py:249]: why do we need to set the height here? And if we do need it, why don't we also set it at line 246, too? 

For the record, the original idea was that these options would be configured at runtime on `run.py`. This way when you deploy a Thalassa instance you could change it according to your needs.

File: thalassa/ui.py
Comment (pmav99) [inline thalassa/ui.py:232]: Please, one assignment per line.

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:137]: 1. Please one assignment per line
2. Perhaps we should consider renaming `x` -> `sx` and `y` -> `sy` (or vice versa) just to be more consistent

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:130]: 1. Please one assignment per line
2. `SCHISM_hgrid_node_x` is hardcoded. This means that the timeseries functionality can only be used with SCHISM output. Trying to use it with anything else will just throw an exception. Please, either pass the names of the variables as arguments to this function, or directly pass the arrays as arguments

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py]: classes should be named using PascalCase: `s/timeseries_data/TimeseriesData/g`

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:143]: I guess that here you are trying to "reset" the `timeseries_data` instance to a "vanilla" state. 

Nevertheless, I don't really like injecting attributes dynamically to "data containers". Principle of least surprise etc. I mean, when you read the code and you encounter the definition of `timeseries_data` class, how will you know that its instances are supposed to have all these extras attributes? how will you know what are their default values? How do you know  that can be reset, too?

These attributes and their default values should be defined in `timeseries_data.__init__()`. Resetting them to their default values could probably be done in `timeseries_data.clear()`.

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:123]: The names of the arguments could probably be improved. E.g. 
- `data` is too generic. How about `ts_data` or something?
- `fmt` is probably some abbreviation. To me is it not apparent which one exactly. Generally speaking, it is usually a good idea not to make too many assumptions WRT who will read our code. Abbreviations, unless their meaning is 100% clear  from the context (e.g. physics formulas), should probably be avoided. Even then their meaning might not be clear to someone who is not a domain expert. To me, out of context, `fmt` can only mean `format`. In this case, if I understand what the code does, it holds the value of the `timeseries_pts` widget and indicates the "mode"  under which we are operating (i.e. `add_points`, `remove_points` or `clear`) which is not related to `format`. I would suggest finding something more specific.
- `source` is only used later on in the holoviews streams and it is a `gv.Trimesh` instance. I think that renaming that to `trimesh` would be both more consistent with the other functions in the `api.py` and would help the reader understand how he should use the `get_timeseries()` function.

File: thalassa/api.py
Comment (pmav99) [inline thalassa/api.py:80]: In this function`data` is an `xr.Dataset` instance.  in the other functions `data` is a `timeseries_data` instance while the `xr.Dataset` instance is called `dataset`. I would suggest keeping the naming consistent.

**Review by pmav99 — CHANGES_REQUESTED** [review]
@wzhengui 
The UI is now much much nicer! Great work!
 
I have a few nitpicks here and there, but nothing really major so I don't think it should take much to wrap this up.

PS. Before each function in `api.py` you have a short comment. Personally I would remove them since they practically repeat the function name (good job on choosing those names BTW). If you want to keep these comments, please convert them to docstrings instead.

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:11]: Hi @pmav99 :  I knew this PEP-8 style, but this one assignment per line is what I really don't like in some cases. Why do people want to make the code long in order to stick to style? Particularly, for some short assignment, why do people prefer several lines instead of one line? Can I keep some of these, which actually makes code compact, easy to maintain, out of my programing experience? If you read a really long code with too much redundancy information, you are actually wasting your memory capacity  (see PEP-8 instruction below). 


A Foolish Consistency is the Hobgoblin of Little Minds

One of Guido's key insights is that code is read much more often than it is written. The guidelines provided here are intended to improve the readability of code and make it consistent across the wide spectrum of Python code. As PEP 20 says, "Readability counts".

A style guide is about consistency. Consistency with this style guide is important. Consistency within a project is more important. Consistency within one module or function is the most important.

However, know when to be inconsistent -- sometimes style guide recommendations just aren't applicable. When in doubt, use your best judgment. Look at other examples and decide what looks best. And don't hesitate to ask!

In particular: do not break backwards compatibility just to comply with this PEP!

Some other good reasons to ignore a particular guideline:

    When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.
    To be consistent with surrounding code that also breaks it (maybe for historic reasons) -- although this is also an opportunity to clean up someone else's mess (in true XP style).
    Because the code in question predates the introduction of the guideline and there is no other reason to be modifying that code.
    When the code needs to remain compatible with older versions of Python that don't support the feature recommended by the style guide.

File: thalassa/ui.py
Comment (wzhengui) [reply thalassa/ui.py:249]: This is to make better use of screen size dynamically when we add or remove a extra windows.

File: thalassa/ui.py
Comment (wzhengui) [reply thalassa/ui.py:232]: See my comment above

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:137]: 1. See my comment above 
2. OK, I will change (x,y) -> (sx,sy) here.

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py]: I don't know what PascalCase is.  I will look into it. Is this a really problem?

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:143]: I will think about this. My original purpose is to make the code efficient and easy to maintain.

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:80]: OK, I will change it later.

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:123]: This is a generic function. Therefore, using generic name is more appropriate.  Later, data can be other instance. Also, source is not limited to Trimesh.

File: thalassa/api.py
Comment (pmav99) [reply thalassa/api.py:11]: Dear @wzhengui thank you for quoting PEP8 and your comments. 

If possible, I'd rather avoid a long discussion about style guides where each one of us will state our personal preferences. This is an open source project where we hope to attract more collaborators in the future and therefore it is important to follow the conventions that most other open source projects are using. In this context, following PEP8 etc is the way to go.

File: thalassa/api.py
Comment (pmav99) [reply thalassa/api.py]: https://www.python.org/dev/peps/pep-0008/#class-names

**Comment by yosoyjay** [comment]
I am of the opinion that using linting for projects to specific conventions aids maintainability and makes it easier for the future user and contributors to understand the code. In this case it looks like the standard is pep8 and black with some deviations specified in the GitHub Actions config file.

@wzhengui How are you currently linting your code?  A simple way to deal with a lot of these comments is to lint your code before you submit a PR.

@yosoyjay : I agree that the coding should follow the general rule of pep8. I am using linting in checking my code, and I don't see any warnings. What I mean in the discussion is that we doesn't go to the extreme of coding styles, but focus on our goals.

**Comment by SorooshMani-NOAA** [comment]
@pmav99, @wzhengui can we set up a style config that can be used by something like autopep8 or black to automatically update the code, then @wzhengui can run the linter on his side, then test the updated code and push again? 
Then the same style config can be used for checking before merge so that there's a established style guide and it doesn't come to personal preferences.

File: thalassa/api.py
Comment (wzhengui) [reply thalassa/api.py:137]: I check the code, and think (x, y) are still better names here because they are local temporary names. Global names should be used carefully.

**Pull request merged**
